binary_search_tree/tree.py

```diff
@@ -14,42 +14,119 @@ class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def add(self, key, value = None):
-        pass
+        new_node = TreeNode(key, value)
+        if not self.root:
+            self.root = new_node
+            return
+        self.add_node(self.root, new_node)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def add_node(self, current, new_node):
+        if current == None:
+            return new_node
+        if  new_node.key <= current.key:
+            current.left = self.add_node(current.left, new_node)
+        else:
+            current.right = self.add_node(current.right, new_node)
+        return current
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)     
     def find(self, key):
```

Comment:
👍 But I'd say the space and time complexities are O(log n) if the tree is balanced and O(n) if unbalanced.